psutil/arch/osx/cpu.c

```diff
@@ -109,7 +114,107 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
     );
 }
 
+#if defined(__arm64__) || defined(__aarch64__)
+PyObject *
+psutil_cpu_freq(PyObject *self, PyObject *args) {
+    uint32_t min;
+    uint32_t curr;
+    uint32_t pMin;
+    uint32_t eMin;
+    uint32_t max;
+    kern_return_t status;
+    CFDictionaryRef matching = NULL;
+    CFTypeRef pCoreRef = NULL;
+    CFTypeRef eCoreRef = NULL;
+    io_iterator_t iter = 0;
+    io_registry_entry_t entry = 0;
+
+    matching = IOServiceMatching("AppleARMIODevice");
+    if (matching == 0) {
+        return PyErr_Format(PyExc_RuntimeError, "IOServiceMatching call failed, 'AppleARMIODevice' not found");
+    }
+
+    status = IOServiceGetMatchingServices(kIOMainPortDefault, matching, &iter);
+    if (status != KERN_SUCCESS) {
+        PyErr_Format(PyExc_RuntimeError, "IOServiceGetMatchingServices call failed");
+        goto error;
+    }
+
+    while ((entry = IOIteratorNext(iter))) {
+        io_name_t name;
+        status = IORegistryEntryGetName(entry, name);
+        if (status != KERN_SUCCESS) {
+            if (entry)
+                IOObjectRelease(entry);
+            continue;
+        }
+        if (strcmp(name, "pmgr") == 0) {
+            break;
+        }
+        IOObjectRelease(entry);
+    }
+
+    if (entry == 0) {
+        PyErr_Format(PyExc_RuntimeError, "'pmgr' entry was not found in AppleARMIODevice service");
+        goto error;
+    }
+
+    pCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states5-sram"), kCFAllocatorDefault, 0);
+    if (!pCoreRef) {
+        PyErr_Format(PyExc_RuntimeError, "'voltage-states5-sram' property not found");
+        goto error;
+    }
+
+    eCoreRef = IORegistryEntryCreateCFProperty(entry, CFSTR("voltage-states1-sram"), kCFAllocatorDefault, 0);
+    if (!eCoreRef) {
+        PyErr_Format(PyExc_RuntimeError, "'voltage-states1-sram' property not found");
+        goto error;
+    }
+
+    size_t pCoreLength = CFDataGetLength(pCoreRef);
+    size_t eCoreLength = CFDataGetLength(eCoreRef);
+    if (pCoreLength < 8) {
+        PyErr_Format(PyExc_RuntimeError, "expected 'voltage-states5-sram' buffer to have at least size 8");
+        goto error;
+    }
+    if (eCoreLength < 4) {
+        PyErr_Format(PyExc_RuntimeError, "expected 'voltage-states1-sram' buffer to have at least size 4");
+        goto error;
+    }
 
+    CFDataGetBytes(pCoreRef, CFRangeMake(0, 4), (UInt8 *) &pMin);
+    CFDataGetBytes(eCoreRef, CFRangeMake(0, 4), (UInt8 *) &eMin);
+    CFDataGetBytes(pCoreRef, CFRangeMake(pCoreLength - 8, 4), (UInt8 *) &max);
+
+    min = pMin < eMin ? pMin : eMin;
+    curr = max;
+
+    if (pCoreRef != NULL)
```

Comment:
I realized you copy-pasted from the error condition, but I'm pretty sure we know this isn't null here (line 163).

Similarly for `eCoreRef` tested on line 160 and `entry` line 157.   And I don't think it'll hurt to call a release on a 0 `iter` (since you already iterated on it).